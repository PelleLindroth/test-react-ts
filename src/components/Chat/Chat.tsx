import { useState } from 'react'
import ChatIcon from '../../assets/chat-icon.png'

interface IMessage {
  text: string,
  id: string
}

const messageHistory: IMessage[] = [
  {
    text: 'Hey how are you',
    id: '345u5y2uy'
  },
  {
    text: 'Never you mind',
    id: 'kjh6jh74567'
  },
  {
    text: 'ok',
    id: 'jhg324jh'
  }
]

const Chat = () => {
  const [isExpanded, setIsExpanded] = useState<boolean>(false)
  const [message, setMessage] = useState<string>('')

  const addMessage = () => {
    messageHistory.push({text: message, id: String(new Date())})
    setMessage('')
  }

  return (
    <>
      <img 
        alt="chat-icon" 
        src={ChatIcon} 
        onClick={() => setIsExpanded(!isExpanded)}
      />
    {
      isExpanded && (
        <>    
         <ul className="message-history">
          {
            messageHistory.map(message => {
              return (
                <li className="message" key={message.id}>{message.text}</li>
              )
            })
          }
         </ul>
         <form onSubmit={addMessage}>
          <label htmlFor="chat-input">Type your message here</label>
          <input onChange={(e) => setMessage(e.target.value)} value={message} id="chat-input" type="text" />
          <button>Send</button>
         </form>
        </>
      )
    }
      </>
  )
}

export default Chat