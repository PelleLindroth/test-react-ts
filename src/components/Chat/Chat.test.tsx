import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { shallow } from "enzyme"
import Chat from './Chat'

describe('Chat', () => {
  it('renders Chat component correctly with chat icon', () => {
    const wrapper = shallow(<Chat />)

    expect(wrapper).toMatchSnapshot()
  })
  it('renders full Chat component correctly when clicking on chat icon', () => {
    const wrapper = shallow(<Chat />)

    wrapper.find('img').simulate('click')

    expect(wrapper).toMatchSnapshot()
  })
  it('shows only icon by default', () => {
    render(<Chat />)

    const icon = screen.getByAltText(/chat-icon/)
    const invisibletextInput = screen.queryByRole('textbox', { name: /message/i })

    expect(icon).toBeInTheDocument()
    expect(invisibletextInput).not.toBeInTheDocument()
  })
  it('expands component when clicking on chat icon', () => {
    render(<Chat />)

    const icon = screen.getByAltText(/chat-icon/)
    const invisibleChatInput = screen.queryByRole('textbox', { name: /message/i })
    expect(invisibleChatInput).not.toBeInTheDocument()
    
    userEvent.click(icon)
    
    const visibleChatInput = screen.getByRole('textbox', { name: /message/i })
    expect(visibleChatInput).toBeInTheDocument()
  })
  it('hides expanded chat when clicking on chat icon again', () => {
    render(<Chat />)

    const icon = screen.getByAltText(/chat-icon/)
    
    userEvent.click(icon)

    const visibleChatInput = screen.getByRole('textbox', { name: /message/i})
    expect(visibleChatInput).toBeInTheDocument()
    
    userEvent.click(icon)

    expect(visibleChatInput).not.toBeInTheDocument()
  })
  it('clears text input on Enter and adds message to message history', () => {
    render(<Chat />)

    const icon = screen.getByAltText(/chat-icon/)
    userEvent.click(icon)
    
    const messages = screen.getAllByRole('listitem')

    expect(messages).toHaveLength(3)

    const textInput = screen.getByRole('textbox', { name: /message/i})
    
    userEvent.type(textInput, 'thats great{enter}')
    
    const updatedMessages = screen.getAllByRole('listitem')

    expect(textInput).toHaveTextContent('')
    expect(updatedMessages).toHaveLength(4)
    expect(updatedMessages[3]).toHaveTextContent('thats great')
  })
})