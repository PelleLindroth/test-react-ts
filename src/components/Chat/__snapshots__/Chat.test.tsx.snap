// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders Chat component correctly with chat icon 1`] = `
<Fragment>
  <img
    alt="chat-icon"
    onClick={[Function]}
    src="chat-icon.png"
  />
</Fragment>
`;

exports[`Chat renders full Chat component correctly when clicking on chat icon 1`] = `
<Fragment>
  <img
    alt="chat-icon"
    onClick={[Function]}
    src="chat-icon.png"
  />
  <ul
    className="message-history"
  >
    <li
      className="message"
      key="345u5y2uy"
    >
      Hey how are you
    </li>
    <li
      className="message"
      key="kjh6jh74567"
    >
      Never you mind
    </li>
    <li
      className="message"
      key="jhg324jh"
    >
      ok
    </li>
  </ul>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="chat-input"
    >
      Type your message here
    </label>
    <input
      id="chat-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button>
      Send
    </button>
  </form>
</Fragment>
`;
